stdlib/SparseArrays/src/sparsematrix.jl

```diff
@@ -192,58 +186,64 @@ function Base.show(io::IO, ::MIME"text/plain", S::AbstractSparseMatrixCSC)
-    function _format_line(r, col, padr, padc)
-        print(ioc, "\n  [", rpad(rowvals(S)[r], padr), ", ", lpad(col, padc), "]  =  ")
-        if isassigned(nonzeros(S), Int(r))
-            show(ioc, nonzeros(S)[r])
+    m, n = size(S)
+    (m == 0 || n == 0) && return show(io, MIME("text/plain"), S)
+    maxHeight = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
+    maxWidth = displaysize(io)[2] ÷ 2
+
+    # Determine if scaling is needed and set the scaling factors accordingly
+    scaleHeight = m
+    scaleWidth = n
+    if get(io, :limit, true) && (m > 4maxHeight || n > 2maxWidth)
+        s = (m * 2maxWidth ÷ n, n * 4maxHeight ÷ m)
+        if s[1] <= 4maxHeight
+            scaleHeight = s[1]
+            scaleWidth = 2maxWidth
         else
-            print(ioc, Base.undef_ref_str)
+            scaleHeight = 4maxHeight
+            scaleWidth = s[2]
         end
     end
 
-    function _get_cols(from, to)
-        idx = eltype(getcolptr(S))[]
-        c = searchsortedlast(getcolptr(S), from)
-        for i = from:to
-            while i == getcolptr(S)[c+1]
-                c +=1
-            end
-            push!(idx, c)
-        end
-        idx
+    brailleBlocks = split("⠁⠂⠄⡀⠈⠐⠠⢀", "") .|> x -> Int(x[1])
+    brailleGrid = fill(10240, (scaleWidth - 1) ÷ 2 + 2, (scaleHeight - 1) ÷ 4 + 1)
+    brailleGrid[end, :] .= 10
+
+    # Given an index `(i, j)` of a matrix `S`, find the corresponding
+    # index `(a, b)` of a matrix of size `height × width`
+    @inline function _scale_index(i::Integer, j::Integer, S::AbstractSparseMatrixCSC, height::Integer, width::Integer)
+        a = size(S, 1) <= 1 || height <= 1 ? 1.0 : (i - 1) / (size(S, 1) - 1) * (height - 1) + 1
+        b = size(S, 2) <= 1 || width <= 1 ? 1.0 : (j - 1) / (size(S, 2) - 1) * (width - 1) + 1
+        return round.(Int, (a, b))
     end
 
-    rows = displaysize(io)[1] - 4 # -4 from [Prompt, header, newline after elements, new prompt]
-    if !get(io, :limit, false) || rows >= nnz(S) # Will the whole matrix fit when printed?
-        cols = _get_cols(1, nnz(S))
-        padr, padc = ndigits.((maximum(rowvals(S)[1:nnz(S)]), cols[end]))
-        _format_line.(1:nnz(S), cols, padr, padc)
-    else
-        if rows <= 2
-            print(io, "\n  \u22ee")
-            return
+    # Given an index `(i, j)` of a matrix, find the corresponding triple
+    # `(k, l, p)` such that the element at `(i, j)` can be found at
+    # position `(k, l)` in the braille grid `brailleGrid` and corresponds
+    # to the 1-dot braille pattern `brailleBlocks[p]`
+    @inline function _to_braille_grid_space(i::Integer, j::Integer)
+        k = (j - 1) ÷ 2 + 1
+        l = (i - 1) ÷ 4 + 1
+        p = ((j - 1) % 2) * 4 + ((i - 1) % 4 + 1)
+        return (k, l, p)
+    end
+
+    @inbounds for j = 1:size(S, 2)
+        for x in nzrange(S, j)
+            si, sj = _scale_index(rowvals(S)[x], j, S, scaleHeight, scaleWidth)
```

Comment:
The repeated call to `rowvals(S)` may be hoisted out by the compiler, but you may as well defined, for instance, `rvals = rowvals(S)` upfront and use it in the loop.